File: 2020/day02/ex2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void assert(const char *buffer, char val, char expected) {
  if (val != expected) {
    fprintf(stderr, "Invalid format: Expected '%c' and got '%c' in '%s'\n", expected, val, buffer);
    exit(1);
  }
}

int main() {
  char buffer[64];
  size_t nb_good = 0;

  while(fgets(buffer, sizeof(buffer), stdin)) {
    char c, pos1_ok, pos2_ok, *idx, *password;
    long pos1, pos2;
    size_t pw_len;

    idx = buffer;
    pos1 = strtol(idx, &idx, 10) - 1;
    assert(buffer, *(idx++), '-');
    pos2 = strtol(idx, &idx, 10) - 1;
    assert(buffer, *(idx++), ' ');
    c = *(idx++);
    assert(buffer, *(idx++), ':');
    assert(buffer, *(idx++), ' ');
    password = idx;
    pw_len = strlen(password);

    pos1_ok = pos1 < pw_len && password[pos1] == c;
    pos2_ok = pos2 < pw_len && password[pos2] == c;

    if (pos1_ok ^ pos2_ok) {
      nb_good++;
    }
  }

  printf("%ld\n", nb_good);

  return 0;
}

File: 2020/day02/ex2.py
import re
import sys

nb_good = 0
for line in sys.stdin:
    m = re.match('(\d+)-(\d+) (\w): (.*)', line)
    if not m:
        raise Exception("No match for %r" % (line,))
    pos1 = int(m.group(1)) - 1
    pos2 = int(m.group(2)) - 1
    charpos1 = m.group(4)[pos1:pos1+1]
    charpos2 = m.group(4)[pos2:pos2+1]
    if ((charpos1 == m.group(3) and charpos2 != m.group(3))
        or (charpos1 != m.group(3) and charpos2 == m.group(3))):
        nb_good += 1
print(nb_good)
